examples/petsc/src/petscutils.c

```diff
@@ -167,3 +176,267 @@ PetscErrorCode CreateRestrictionFromPlex(Ceed ceed, DM dm, CeedInt height,
+
+        interp[q * P + p_ceed] = basis_tabulation->T[0][((face * Q + q) * P * num_comp +
+                                 p_petsc) * num_comp + c];
+        for (CeedInt d = 0; d < dim; d++) {
+          grad[(d * Q + q) * P + p_ceed] = basis_tabulation->T[1][(((
+                                             face * Q + q) * P * num_comp + p_petsc) * num_comp + c) * dim + d];
+        }
+      }
+    }
+
+    // -- Cleanup
+    if (permutation) PetscCall(ISRestoreIndices(permutation, &permutation_indices));
+    PetscCall(ISDestroy(&permutation));
+  }
+
+  // Finally, create libCEED basis
+  CeedBasisCreateH1(ceed, elem_topo, num_comp, P, Q, interp, grad, q_points,
+                    q_weights, basis);
+  PetscCall(PetscFree(q_points));
+  PetscCall(PetscFree(interp));
+  PetscCall(PetscFree(grad));
+
+  PetscFunctionReturn(0);
+}
+
+// -----------------------------------------------------------------------------
+// Get CEED Basis from DMPlex
+// -----------------------------------------------------------------------------
+PetscErrorCode CreateBasisFromPlex(Ceed ceed, DM dm, DMLabel domain_label,
+                                   CeedInt label_value, CeedInt height,
+                                   CeedInt dm_field, CeedBasis *basis) {
+  PetscDS         ds;
+  PetscFE         fe;
+  PetscQuadrature quadrature;
+  PetscBool       is_simplex = PETSC_TRUE;
+  PetscInt        ds_field   = -1;
+
+  PetscFunctionBeginUser;
+
+  // Get element information
+  PetscCall(DMGetRegionDS(dm, domain_label, NULL, &ds));
+  PetscCall(DMFieldToDSField(dm, domain_label, dm_field, &ds_field));
+  PetscCall(PetscDSGetDiscretization(ds, ds_field, (PetscObject *)&fe));
+  PetscCall(PetscFEGetHeightSubspace(fe, height, &fe));
+  PetscCall(PetscFEGetQuadrature(fe, &quadrature));
+
+  // Check if simplex or tensor-product mesh
+  PetscCall(DMPlexIsSimplex(dm, &is_simplex));
+
+  // Build libCEED basis
+  if (is_simplex) {
+    PetscTabulation basis_tabulation;
+    PetscInt        num_derivatives = 1, face = 0;
+
+    PetscCall(PetscFEGetCellTabulation(fe, num_derivatives, &basis_tabulation));
+    PetscCall(BasisCreateFromTabulation(ceed, dm, domain_label, label_value, height,
+                                        face, fe, basis_tabulation, quadrature, basis));
+  } else {
+    PetscDualSpace dual_space;
+    PetscInt       num_dual_basis_vectors;
+    PetscInt       dim, num_comp, P, Q;
+
+    PetscCall(PetscFEGetSpatialDimension(fe, &dim));
+    PetscCall(PetscFEGetNumComponents(fe, &num_comp));
+    PetscCall(PetscFEGetDualSpace(fe, &dual_space));
+    PetscCall(PetscDualSpaceGetDimension(dual_space, &num_dual_basis_vectors));
+    P = num_dual_basis_vectors / num_comp;
+    PetscCall(PetscQuadratureGetData(quadrature, NULL, NULL, &Q, NULL, NULL));
+
+    CeedInt P_1d = (CeedInt)round(pow(P, 1.0 / dim));
+    CeedInt Q_1d = (CeedInt)round(pow(Q, 1.0 / dim));
+
+    CeedBasisCreateTensorH1Lagrange(ceed, dim, num_comp, P_1d, Q_1d, CEED_GAUSS,
```

Comment:
Hard-coding this here breaks bp5/6

Reply:
Hmm, For simplex we use the tabulation. This is for simplex
```
./bps -problem bp5 -degree 3 -ksp_max_it 1000 -simplex

-- CEED Benchmark Problem 5 -- libCEED + PETSc --
  MPI:
    Hostname                                : rezgar
    Total ranks                             : 1
    Ranks per compute node                  : 1
  PETSc:
    PETSc Vec Type                          : seq
  libCEED:
    libCEED Backend                         : /cpu/self/avx/blocked
    libCEED Backend MemType                 : host
  Mesh:
    Solution Order (P)                      : 4
    Quadrature  Order (Q)                   : 4
    Additional quadrature points (q_extra)  : 0
    Global nodes                            : 512
    Local Elements                          : 162
    Element topology                        : tetrahedron
    Owned nodes                             : 512
    DoF per node                            : 1
  KSP:
    KSP Type                                : cg
    KSP Convergence                         : CONVERGED_RTOL
    Total KSP Iterations                    : 49
    Final rnorm                             : 1.956289e-10
  Performance:
    Pointwise Error (max)                   : 1.277850e-01
    CG Solve Time                           : 0.03456 (0.03456) sec
    DoFs/Sec in CG                          : 0.725925 (0.725925) million
```

Reply:
bp5/6 is for collocated quadrature - that doesn't make sense with simplex elements, and currently there is no collocated quadrature for the hex elements with bp5/6

Reply:
Ah, I forgot that. thanks.

Reply:
Note, open ToDo here

Reply:
What we should do about here?

Reply:
I don't see an obvious best path. Somehow 

1) we should fail (preferably early) for simplex + BP5/6 

2) we need to make available in this function the information about which quadrature to use for tensor product elements 

1) feels easy. 2) has many options and I'm not sure which is cleanest/clearer. Maybe an additional function arg?

Reply:
I added bp_data, so we can change the q_mode for hex element. We can add a statement that if `bp_choice = CEED_BP4` or `bp_choice = CEED_BP5` and `is_simplex` break the run, but I don't know how to add that statement such that breaks the execution.